fsharp-backend/src/LibBackend/EventQueueV2.fs

```diff
@@ -0,0 +1,392 @@
+#nowarn "44"
+
+let publisher : Lazy<Task<PublisherServiceApiClient>> =
+  lazy
+    (task {
+      let! client =
+        match credentials with
+        | None ->
+          PublisherServiceApiClientBuilder(
+            EmulatorDetection = Google.Api.Gax.EmulatorDetection.EmulatorOrProduction
+          )
+            .BuildAsync()
+        | Some credentials ->
+          task {
+            let endpoint = PublisherServiceApiClient.DefaultEndpoint
+            let channel = Grpc.Core.Channel(endpoint, credentials)
+            let grpcClient = Publisher.PublisherClient(channel)
+            let settings = PublisherServiceApiSettings()
+            return PublisherServiceApiClientImpl(grpcClient, settings)
+          }
+
+
+      // Ensure the topic is created locally
+      if Config.queuePubSubCreateTopic then
+        let! topicFound =
+          task {
+            try
+              let! _ = client.GetTopicAsync(topicName)
+              return true
+            with
+            | _ -> return false
+          }
+        if not topicFound then
+          let! (_ : Topic) = client.CreateTopicAsync(topicName)
+          ()
+
+
+      return client
+    })
+
+
+
+let subscriber : Lazy<Task<SubscriberServiceApiClient>> =
+  lazy
+    (task {
+      let! (_ : PublisherServiceApiClient) = publisher.Force()
+
+      // Ensure subscription is created locally
+      let! client =
+        match credentials with
+        | None ->
+          SubscriberServiceApiClientBuilder(
+            EmulatorDetection = Google.Api.Gax.EmulatorDetection.EmulatorOrProduction
+          )
+            .BuildAsync()
+        | Some credentials ->
+          task {
+            let endpoint = SubscriberServiceApiClient.DefaultEndpoint
+            let channel = Grpc.Core.Channel(endpoint, credentials)
+            let grpcClient = Subscriber.SubscriberClient(channel)
+            let settings = SubscriberServiceApiSettings()
+            return SubscriberServiceApiClientImpl(grpcClient, settings)
+          }
+
+
+      if Config.queuePubSubCreateTopic then
+        let! subFound =
+          task {
+            try
+              let! _ = client.GetSubscriptionAsync(subscriptionName)
+              return true
+            with
+            | _ -> return false
+          }
+        if not subFound then
+          let! (_ : Subscription) =
+            client.CreateSubscriptionAsync(
+              subscriptionName,
+              topicName,
+              pushConfig = null,
+              ackDeadlineSeconds = 60
+            )
+          ()
+      return client
+    })
+
+let dequeue () : Task<Notification> =
+  task {
+    let! subscriber = subscriber.Force()
+    let! response = subscriber.PullAsync(subscriptionName, maxMessages = 1)
+    let mutable notification : Option<Notification> = None
+    while notification = None do
+      // Messages is allowed return no messages. It will wait a while by default
```

Comment:
Do I have this right:
the result of `subscriber.PullAsync` includes a `ReceivedMessages` that _they_ set, and we're looping around until it's non-empty?

If so, likely worth adding that small explanation here.

By "It will wait a while by default" do you mean that we shouldn't expect a non-empty set for the first few iterations?

Reply:
Yes, that's right. Will update